libs/datasets/sources/covid_county_data.py

```diff
@@ -40,10 +44,88 @@ class CovidCountyDataDataSource(data_source.DataSource):
         }
     }
 
+    @classmethod
+    def synthesize_test_metrics(cls, data: pd.DataFrame) -> Tuple[pd.DataFrame, pd.Series]:
+        """
+        Synthesize testing metrics where they can be calculated from other metrics.
+
+        This function processes each date independently. The fix applied depends on what testing and cases
+        metrics have a real value on that date. The following table lists the number of FIPS-date rows with
+        a given combination of metrics and the treatment applied by this function. The table was generated
+        with:
+
+        columns_to_check = ["negative_tests", "positive_tests", "total_tests", "cases"]
+        df[columns_to_check].notna().groupby(columns_to_check).size().reset_index(name="count").sort_values("count")
+
+        negative_tests  positive_tests  total_tests  cases  count    Treatment / reason for doing nothing
+        True            False           False        True   6        Very rare
+        False           False           True         False  39       Very rare
+        True            True            True         False  245      Testing data already set
+        False           False           False        False  361      No data available
+        True            False           True         True   627      pos = tests - neg
+        False           True            False        True   1079     Testing data not available
+        True            True            False        True   7064     Pos and neg tests set, total_tests not used
+        False           True            True         True   25769    neg = tests - pos
+        False           False           True         True   32684    pos = cases, neg = tests - pos
+        True            True            True         True   61951    Have everything already
+        False           False           False        True   508238   No test data
+        """
+
+        # Make a copy to avoid modifying the argument when using mask with inplace=True.
+        df = data.copy()
+        tests_and_cases = df.eval(
+            "negative_tests.isna() & positive_tests.isna() & total_tests.notna() & cases.notna()"
+        )
+        missing_neg = df.eval(
+            "negative_tests.isna() & positive_tests.notna() & total_tests.notna() & cases.notna()"
+        )
+        missing_pos = df.eval(
+            "negative_tests.notna() & positive_tests.isna() & total_tests.notna() & cases.notna()"
+        )
+
+        # Keep the same order of rows in `provenance` as `df` so that the same masks can be used when
+        # writing it.
+        provenance = df.loc[:, [CommonFields.FIPS, CommonFields.DATE]]
+        provenance[CommonFields.POSITIVE_TESTS] = "none"
```

Comment:
Might be nice to have the different provenance cases as a class:
```
class TestingProvenance:
    NONE = "none"
    TESTS_AND_CASES = "tests_and_cases"
    ...

```

Reply:
I'm not at all sure we want to stick with this, for now think of these values as more like logging comments that ride along with the data.